CLI v2

bentsherman opened the pull request:
Closes #3595 

To summarize:

- `nextflow.cli.*Impl` classes implement the core functionality of each command
- `nextflow.cli.v1.*Cmd` classes implement the CLI v1 and pass options to corresponding `*Impl` class
- `nextflow.cli.v2.*Cmd` classes implement the CLI v2 and pass options to corresponding `*Impl` class
- `nextflow.cli.v1.Launcher` and `nextflow.cli.v2.Launcher` provide entrypoints to the v1 and v2 CLI respectively

TODO:
- [x] Add basic v2 CLI
- [x] Update tests
- [ ] ~Test support for list values~ (keep same as CLI v1)
- [x] Test support for params
- [ ] ~Test support for basic type conversion~ (keep same as CLI v1)
- [x] Update build/install process to provide both CLIs

bentsherman committed:
wip [ci skip]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

pditommaso commented:
I'd propose keeping the current CLI in the current package. That would make the PR much simpler

bentsherman committed:
Update tests

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Remove `kuberun` command from v2 CLI

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
@pditommaso any suggestions for how to change the build process to create both CLIs?

bentsherman committed:
minor edits [ci skip]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add proper params parsing to CLI v2 [ci fast]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

pditommaso commented:
What's your need?

bentsherman commented:
How do you recommend we change the build process to create both the `nextflow` and `nf` binaries?

bentsherman committed:
Update params parsing in CLI v2 to match v1 [ci fast]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add fallback values to CLI options, fix some error messages [ci fast]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Minor improvements

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
@pditommaso Today I realized that picocli can't really tell the difference between pipeline args and pipeline params in the `run` command. Consider the example:
```bash
nextflow run script.nf arg1 arg2 --param1 value1 --param2=value2
```

Whereas jcommander can use `@DynamicParameter(names='--')` to catch the params, picocli doesn't have the same trick. So picocli will interpret all of these as args, regardless of whether they occur before or after the double dash `--`. The double dash is only useful to distinguish params from CLI options.

So there are two solutions:

1. Remove support for pipeline args. It's a breaking change but I'm not sure that anyone uses positional args on pipeline scripts in the first place?
2. Assume the first positional arg that starts with `--` is the first param and interpret everything after that as params.

I'm assuming you don't want to do (1), so let's explore (2) for a moment.

In the above example, picocli would parse the positional args as:
```groovy
args = ['arg1', 'arg2', '--param1', 'value1', '--param2=value2']
```

Then Nextflow would separate the args from params as follows:
```groovy
scriptArgs = ['arg1', 'arg2']
args = ['param1': 'value1', 'param2': 'value2']
```

Since the CLI has no idea what args or params are expected by the pipeline script at this point, I think this is the best we can do. Basically it means that you can't provide an arg value that looks like an option.

bentsherman committed:
Fix issue with multiple positional args, update tests [ci fast]

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Revert build files to use v1 CLI, add launch-v2.sh to test v2 CLI

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Minor edits

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix bug with plugin command

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix several bugs and inconsistencies

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix failing tests

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Merge branch '3595-nf-cli' of github.com:nextflow-io/nextflow into 3595-nf-cli

pditommaso committed:
Merge branch 'master' into 3595-nf-cli

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/cli/CleanImpl.groovy:
```diff
@@ -41,58 +40,39 @@ import nextflow.trace.TraceRecord
-
-    static final public NAME = 'clean'
-
-    @Parameter(names=['-q', '-quiet'], description = 'Do not print names of files removed', arity = 0)
-    boolean quiet
-
-    @Parameter(names=['-f', '-force'], description = 'Force clean command', arity = 0)
-    boolean force
-
-    @Parameter(names=['-n', '-dry-run'], description = 'Print names of file to be removed without deleting them' , arity = 0)
-    boolean dryRun
-
-    @Parameter(names='-after', description = 'Clean up runs executed after the specified one')
-    String after
-
-    @Parameter(names='-before', description = 'Clean up runs executed before the specified one')
-    String before
-
-    @Parameter(names='-but', description = 'Clean up all runs except the specified one')
-    String but
-
-    @Parameter(names=['-k', '-keep-logs'], description = 'Removes only temporary files but retains execution log entries and metadata')
```
Looks smart, but not getting where these are implemented

bentsherman replied:
They are implemented in `cli.v1.CleanCmd` and `cli.v2.CleanCmd` respectively.

pditommaso commented:
@bentsherman some tests area failing. Can you please have a look to it?

bentsherman commented:
There were some tests I forgot to update, most of them only run on Github so I didn't notice. Should be fine now...

pditommaso commented:
A few tests I've done 

```
» ./launch-v2.sh --log hello.log run hello  
N E X T F L O W  ~  version 23.02.1-edge
Cannot invoke "java.lang.Iterable.iterator()" because "self" is null
```

Also, it looks using the an argument with a value as the name of an option cause a conflict, for example: 

```
» ./launch-v2.sh --log log run hello  
Expected parameter for option '--log' but found 'log'
```

bentsherman committed:
Fix several parsing errors

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Allow subcommands to be specified as option values

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
I fixed both of those errors. By default, picocli does not allow subcommands to be specified as option values, so I changed it to be allowed. It should be fine as long as none of the launcher CLI options have an unbounded size.

For example, if the `--trace` option allowed multiple values separated by spaces, and it was specified before the subcommand, then it would consume the subcommand as a value. But since `--trace` expects values separate by a comma, it won't have this problem. In any case, we will have to be careful about how we design CLI options if we keep this setting.

If subcommands aren't allowed as option values, you can also escape the value like so:
```bash
./launch-v2.sh --log=\"log\" run hello  
```

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

pditommaso committed:
Merge branch 'master' into 3595-nf-cli

bentsherman commented on launch-v2.sh:
```diff
@@ -0,0 +1,17 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+MAIN_CLASS="nextflow.cli.v2.Launcher" ./launch.sh $@
```
@sonatype-lift ignore

bentsherman commented on modules/nextflow/src/main/groovy/nextflow/cli/FsImpl.groovy:
```diff
@@ -0,0 +1,116 @@
+    void cat(Path source) {
+        String line
+        def reader = Files.newBufferedReader(source, Charset.defaultCharset())
+        while( line = reader.readLine() )
```
@sonatype-lift ignore

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add `nf` binary

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix issue with ansi logger

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
The CLI v2 can now be used like a regular command:
```bash
# install nextflow JAR from source
make install

# add nf command to $PATH
cp nf ~/.local/bin

# try it!
nf run hello
```

bentsherman committed:
Cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Put CLI v1 code back into original command classes

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Update tests

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
@pditommaso This PR is ready for review. It is significantly smaller and should be easier to review, just make sure "Hide whitespace" is enabled in the diff viewer.

The `CmdPullTest` is failing but not sure why since it uses the Github access token.

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix inspect command

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
I think we should have some test suite for CLI commands. Here's what I have so far:

```bash
cmd='../launch-v2.sh'
pipeline='rnaseq-nf'

$cmd
$cmd --version
$cmd help
$cmd info

$cmd pull $pipeline
$cmd list
$cmd view $pipeline
$cmd config $pipeline

$cmd inspect $pipeline

$cmd run $pipeline --with-conda

$cmd log
$cmd log last

$cmd clean --force

$cmd drop $pipeline
```

bentsherman committed:
Fix failing test

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Rename `--deep` to `--depth` to match git

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Remove `-dsl1` and `-dsl2` options

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Remove `-without-*` options from CLI v2

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Update docs

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add nf to install procedure

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
The `nf` command should now be created automatically when installing Nextflow from `get.nextflow.io` or upgrading via `nextflow self-update`

marcodelapierre commented:
Hi @bentsherman , I wanted to test but I am getting a build time error, i.e. when running `make`:

```
./gradlew clean

FAILURE: Build failed with an exception.

* What went wrong:
Could not open settings generic class cache for settings file '/home/ubuntu/develop/test-cli2/nextflow/settings.gradle' (/home/ubuntu/.gradle/caches/7.6.1/scripts/3vrv8lwkdc8liai4nlcjtgty9).
> BUG! exception in phase 'semantic analysis' in source unit '_BuildScript_' Unsupported class file major version 65
```

Note how it is using Gradle 7.6.1 (current master uses 8.x).

I am running on Ubuntu with GraalVM CE 21+35.1.

Does it build for you?

marcodelapierre commented:
@bentsherman  never mind I responded to myself while writing the comment above ... I downgraded to Java 17 and now I can build ;)

marcodelapierre commented:
So, I have done some basic testing, along the lines of:

```
../nextflow/launch-v1.sh run nextflow-io/hello -with-docker

../nextflow/launch-v2.sh run nextflow-io/hello --with-docker
```

Two key reports so far:

1. if `v2` is used with `-with-docker` (single dash), it will silently ignore it, as far as the option,`with-docker` in this case, is legal (e.g.  `-blue-moon` will fail and `--blue-moon` will both fail). Is this expected? I would say it should throw an error.

2. if `v1 is used with `--with-docker` (double dash), it will still go ahead and execute. I assume it is considering it as a pipeline parameter (i.e. as if there were a `params.with-docker` in the pipeline)

marcodelapierre commented:
Now for the use of pipeline parameters:

```
../nextflow/launch-v1.sh run marcodelapierre/toy-cowsay-nf -r main -profile docker

../nextflow/launch-v2.sh run marcodelapierre/toy-cowsay-nf -r main --profile docker
```

What I notice with `v2` is that I always use the double dash separator for the input parameters.

This errors:
```
../nextflow/launch-v2.sh run marcodelapierre/toy-cowsay-nf -r main --profile docker --hello=Kaya,Hello
```

This runs:
```
../nextflow/launch-v2.sh run marcodelapierre/toy-cowsay-nf -r main --profile docker -- --hello=Kaya,Hello
```

To be honest, I like this behaviour. With the double dash both specifying runtime options and pipeline parameters, we need a way to handle the fact that the latter can take any name, and so:
1. runtime options with unknown names must error
2. pipeline parameters have to able to take any naeme

In this context, the compulsory double dash seems effective to me in maintaining the two sets of options distinct.

I am reflecting on this as my initial undertsanding (from reading the docs in this PR) was that the separator double dash was only needed when the pipeline parameter has a name that conflicts with a runtime parameter.


Other than reviewing the docs to make sure they are stating the correct behaviour, I am OK with this behaviour.

@pditommaso @bentsherman  ?

marcodelapierre commented:
On the new `nf` command, possibly a naïve question, is there a way to test it right now, without resorting to `launch-v2.sh`? 

Normally for v1 I would be able to use `launch-v1.sh` and `nextflow` interchangeably with local builds.

marcodelapierre reviewed (changes requested):
See my comment in the `cli.md` docs file, 

plus these two in the main thread: 

* https://github.com/nextflow-io/nextflow/pull/3600#issuecomment-1815904236 (point 1.)
* https://github.com/nextflow-io/nextflow/pull/3600#issuecomment-1815940653

marcodelapierre commented:
Finally, @bentsherman is there any specific point of focus you want me to focus on in the codebase?
So far I have left out of the code review all of the files under `modules/nextflow/src/main/groovy/nextflow/cli/` and all of the unit tests.

bentsherman commented:
Thanks @marcodelapierre for your comments. I think you should focus on the testing as you have been doing, to find any quirks or things that could be improved with the user experience.

- The CLI v1 does not check the params for mistakes like `-resume -> --resume`, but it could be useful. Could be done in a separate PR as it doesn't depend on CLI v2.
- The CLI v2 does not check for old single-dash options. I think they would be parsed as args. Should be easy to check though. As with CLI v1, it is unlikely that someone would intentionally define a param with the same name as a CLI option, so it's probably worthwhile to always warn the user.
- CLI v2 params should work even without the double dash. Maybe the problem is the equals sign?
- If you run `make install ; cp nextflow $(which nextflow) ; nextflow self-update`, it should install the local build and `nf` CLI. Just keep in mind it will overwrite your current nextflow installation.

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix build errors

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix failing tests

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

marcodelapierre commented:
@bentsherman please let me know when you are done solving the conflicts from master, and if you need any specific check on that

bentsherman commented:
@marcodelapierre should be ready to test now

marcodelapierre commented:
*** ignore until working day ***

minor makeup, when running `nf` without parameters:
```
[...]

  run          Execute a pipeline
  self-update  Update nextflow runtime to the latest available version
  view         View project script file(s)
  secrets      Manage pipeline secrets (preview)
```

`secrets` comes last, not following alphabetic sorting. I know this item is added conditionally, so would it be possible to sort these items before outputting?

marcodelapierre commented:
Some comments by looking at the output of `nf run -h` (some are just face lifting, others are around implementation).

Is the first equal desired in these two instances? (I say we should have either the dot or the equal, not both, with the equal probably being more Linuxesque)
- `-e., --env.=<name>=<value>` 
- --process.[=<name>[=<value>]]
Also, I don't think we need the outer square brackets for the `process` option above.

`--disable-jobs-cancellation` : I would suggest `--jobs-cancellation=true|false`

`--dump-hashes=<dumpHashes>` : CLI v1 had without values would dump all hashes by default; what is the intended behaviour with v2? Also, I tried specifying a task name, and it would still dump hashes for ALL tasks (`launch-v2.sh run rnaseq-nf --with-docker --dump-hashes MULTIQC`)

`--resume[=<resume>]` : what is the meaning of the optional value ? - should be described in the help description

marcodelapierre commented:
> The CLI v2 does not check for old single-dash options. I think they would be parsed as args. Should be easy to check though. As with CLI v1, it is unlikely that someone would intentionally define a param with the same name as a CLI option, so it's probably worthwhile to always warn the user. 

Maybe both CLI v1 and v2 could check for unused parsed args, and print a warning about them?  I am shooting a separated issue, as this is unrelated to V2: https://github.com/nextflow-io/nextflow/issues/4567

marcodelapierre commented:
My main report is about handling of arguments, vs using or not a double-dash separator.

Baseline command: `launch-v2.sh run hello -r add/params.hello`.  
Same behaviours observed also when git cloning the pipeline and running from its directory, `launch-v2.sh run main.nf`.   

1. unused spurious arguments, e.g. `bluemoon` or `--  bluemoon`: silently ignored - OK

2. double dash prefixed arguments (in principle could be pipeline params)
  A. no separator double dash and actual pipeline parameter, e.g. `--hello Kaya` or `--hello=Kaya`:  `Unknown option` error
  B. no separator double dash and spurious, e.g. `--bluemoon` or `--blue moon` or `--blue=moon`: `Unknown option` error
  C. separator double dash and actual pipeline parameter, e.g. `--   --hello Kaya` or `--   --hello=Kaya`:  used - OK
  D. separator double dash and spurious, e.g. `--   --bluemoon` or `--   --blue moon` or `--   --blue=moon`: silently ignored - OK

3. single dash prefixed arguments (in the style of CLI v1 runtime options)
  A. no separator double dash and existing in CLI v1, e.g. `-with-docker`or `-with-docker crocodile` or `-with-docker=crocodile`: silently ignored
  B. no separator double dash and not existing in CLI v2, e.g. `-bluemoon` or `-blue moon` or `-blue=moon`:  `Unknown option` error
  C. separator double dash and existing in CLI v1, e.g. `--  -with-docker`or `--  -with-docker crocodile` or `--  -with-docker=crocodile`: silently ignored - OK
  D. separator double dash and not existing in CLI v2, e.g. `--  -bluemoon` or `--  -blue moon` or `--  -blue=moon`: silently ignored - OK


Two key comments for you @bentsherman .

I keep seeing an unintended behaviour around using or not using a double dash separator to provide arguments, unlike what you describe. In particular, omitting the double dash separator results in undesired behaviours, see above points 2.A, 2.B and 3.B. What are we doing differently from each other?  How can we make the non double dash separated syntax to work as expected?   
NOTE: I am using this branch, then `make compile ; make install`, then using `launch-v2.sh`.  

I notice that V2 still preserves memory of V1 runtime options, at least when non using the double dash separator. In fact, see points above, with the current code status truly unknown options throw an error (2.A, 2.B and 3.B); however case 3.A, where the single dashed V1 options are ignored rather than throwing an error. How can we make V2 to not have any knowledge about the legacy single dash prefixed options? (this is unnecessary for V2)

marcodelapierre commented:
I am done with my testing, so Ben once you address the points above, I will be happy with merging

marcodelapierre commented:
I am done with my testing, so Ben once you address the points above, I will be happy with merging

ewels commented:
It'd be great if the CLI could exit immediately with an error if the user tries to provide a parameter that matches a core option name.

eg. `Core options not allowed as parameters. Did you mean..?`

marcodelapierre commented:
> It'd be great if the CLI could exit immediately with an error if the user tries to provide a parameter that matches a core option name.

Good point.   @bentsherman after the other points above are fixed, let's see whether to include this functionality with this PR, or in a separate subsequent one.

bentsherman committed:
Fix bugs

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Minor edits

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
FIx inspect command

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Fix help text for run options

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
Okay I have addressed a number of issues:

- the `secrets` command was being added conditionally because it used to require an environment variable `NXF_ENABLE_SECRETS` to be enabled, but now it is enabled by default so we can just add it in the main list

- the ugly exception you got when you ran a pipeline that would fail was just a bug on my end, it is fixed

- run command wasn't creating the log file, that is fixed

- not sure why the params stopped working without the double dash, but I have fixed it

- CLI v2 will log a warning if it detects a single-dash option while parsing params, with or without the `--` separator

- however, the reason why `-with-docker` isn't caught is because it is interpreted as `-w ith-docker`, as in "set the work directory to ith-docker" 🤦 The same problem will happen for any v1 option that would match a v2 short option with value.

  We could [disable clustered short options](https://picocli.info/#_posix_clustered_short_options) so that `-with-docker` is not parsed as `-w ith-docker`, but other options like `-process.<name>=<value>` depend on this syntax.

  Or, we might be able to implement a [custom parser](https://picocli.info/#_custom_parameter_processing) that simply checks for v1 options ahead of time and then defers to the default parser.

- There is no way to distinguish between v2 options and pipeline params for the purpose of raising an error. If you don't use the `--` separator, the param will simply be interpreted as a CLI option. Whereas CLI v1 could do this since its options are single-dash, for CLI v2 we will need to depend on the functionality of nf-validation, which checks the params for name conflicts after the fact.

bentsherman commented:
Also Marco regarding your earlier questions about some CLI options:

- Key-value options like `--env`, `--executor`, `--process`, etc should be specified as `--process.<name>[=<value>]` (the `=` is optional). I tried to use `paramLabel` to show this in the help text, but unfortunately picocli inserts an equals sign so that it becomes `--process.=<name>=<value>`. Not sure how to fix this without customizing the entire help text.

  Related to my previous comment, this syntax depends on the use of "clustered short options" like `-Dkey=value`. I have always considered this a weird hack, and considering the above issue with v1 options like `-with-docker`, maybe we should just disable the clustered short options and require something like `--process <name>=<value>` (now the `=` is required).

- `--dump-hashes` can have no value, `default`, or `json`, which determines the output format of the task hashes. It is documented in more detail in the docs, I agree the CLI option should also be more explicit, but I think it would be better to do a separate PR in which we align the CLI help text with the docs. This PR is big enough so trying to keep it focused on the addition of CLI v2.

- `--resume` can have no value, `last` (default), or a session ID from which to resume. So by default it resumes from the session ID of the last run you did. Same as above, we should improve this in a separate PR.

bentsherman commented:
In the meantime I was able to improve the help text for these options through the param label, which CLI v1 doesn't have. For the key-value options I did something like this:
```
--process.[=*]         Set process options (e.g. --process.<name>=<value>)
```

Still a little ugly but the `[=]` can't be disabled AFAIK. I would still support changing these options to be `--process <key>=<value>` and disabling the clustered short options, but know that it would be a "breaking" change from v1 beyond the simple rule of adding a second dash to every option.. Of course I have already made a few such changes and documented them in the CLI docs, so could just be another item on that list.

bentsherman committed:
Fix merge conflicts

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Disable clustered short options

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Make v1 option warning yellow

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

marcodelapierre commented:
> Related to my previous comment, this syntax depends on the use of "clustered short options" like -Dkey=value. I have always considered this a weird hack, and considering the above issue with v1 options like -with-docker, maybe we should just disable the clustered short options and require something like --process <name>=<value> (now the = is required). 

I think we should strongly consider disable the clustered short options within in this PR. I see a thing like `-with-docker` used in v2 as a possible or unexplicable errors and frustration for the users. As we are moving to CLI v2, it might be a good time  to push for a change towards `--process <name>=<value>` now.  

@bentsherman @pditommaso thoughts?

marcodelapierre commented:
> CLI v2 will log a warning if it detects a single-dash option while parsing params, with or without the -- separator

Amazing!   @ewels  FYI

marcodelapierre reviewed (approved):
I think we should strongly consider disable the clustered short options within in this PR. 

See two comments above

bentsherman committed:
Merge branch 'master' into 3595-nf-cli

ewels commented:
Agree, the syntax with the extra space looks fine to me so not sure that the clustered short thing adds a lot. 

> maybe we should just disable the clustered short options and require something like `--process <name>=<value>` (now the = is required).

I think this would be fine 👆🏻

marcodelapierre committed:
Merge branch 'master' into 3595-nf-cli

bentsherman committed:
Fix map options in run command

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
I agree, okay I disabled the clustered short options and made the warning yellow. The ANSI logger isn't initialized at this point so I just used the ANSI codes directly

bentsherman commented:
In order to support pipeline params alongside CLI options and detect spurious v1 options in the run command, we basically have to disable the standard error checking for that command, i.e. to say "missing options ..." if an invalid option is provided.

To fully recover this error checking we will need some way to validate CLI params against the params defined by the pipeline. This is tricky because params can be defined across multiple configs and even in the pipeline itself. But I think we could make it work:
- Parse CLI options, eagerly accept unmatched options as params
- Parse config files, which produces a full set of params
- Warn the user if a CLI "param" was not defined by the config file
- Ignore params defined in the pipeline, they can't be known until running the script and they should still be defined in the config file for posterity

I think this is a separate effort from CLI v2, because it would also be useful for CLI v1.

marcodelapierre commented:
> In order to support pipeline params alongside CLI options and detect spurious v1 options in the run command, we basically have to disable the standard error checking for that command, i.e. to say "missing options ..." if an invalid option is provided.
> 
> To fully recover this error checking we will need some way to validate CLI params against the params defined by the pipeline. This is tricky because params can be defined across multiple configs and even in the pipeline itself. But I think we could make it work:
> 
> * Parse CLI options, eagerly accept unmatched options as params
> * Parse config files, which produces a full set of params
> * Warn the user if a CLI "param" was not defined by the config file
> * Ignore params defined in the pipeline, they can't be known until running the script and they should still be defined in the config file for posterity
> 
> I think this is a separate effort from CLI v2, because it would also be useful for CLI v1.

Agree, separate issue

marcodelapierre commented:
I have re-tested, all good with the clustering disabled!

Super duper minimal last thing, if possible:  in the `nextflow run` help screen, the `process` line,  `--process[=<name>=<value>]`, should not have the square brackets.


Otherwise let's warm up @pditommaso  for the merge!  
Pending:
- Last CI checks successful
- @bentsherman  considering, and eventually addressing, the minor glitch in the help screen, see above in this comment

marcodelapierre commented:
Great job Ben!

ewels commented:
> The ANSI logger isn't initialized at this point so I just used the ANSI codes directly

This is a bit risky - note that the ANSI codes will be dumped to redirected text files / anything that can't support ANSI output.

The ANSI logger is smarter than that and strips the codes if the output doesn't support them.

ewels commented:
Couple of thoughts from looking at the help output:

* Any reason why `-remote-debug` only has one hyphen?
* Some options like `--bg` look like short versions, could that be `-bg`/`--background`?
* Can we print the Nextflow 1-line header above help text? (Maybe say, with a [nice green background](https://github.com/nextflow-io/nextflow/pull/4573)? - would require ANSI library init though)

marcodelapierre commented:
@pditommaso , we would be keen on your feedback on this PR, firstly from a feature-set perspective, and optionally from the tech standpoint. Thank you!

bentsherman committed:
Remove ANSI shim from params helper

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Add --background to run command

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Minor edits

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

marcodelapierre commented:
@bentsherman can you check the merge conflicts from master?

bentsherman commented:
For the new CLI I think we should consider removing support for args in the run command. You can specify args and access them in the pipeline script via `args[0]`, `args[1]`, etc. This seems like more of a DSL1 concept that doesn't really fit with our vision for formalizing the workflow inputs. And I've never seen anyone use it, not sure if anyone even knows it exists. Removing it would simplify the params parsing in the new CLI. Just some food for thought for the reviewers.

marcodelapierre commented:
Good point Ben. 

I personally did NOT know of the feature. I see it is not documented in docs.nextflow.io, and that it is not used across nf-core (see e.g. this search https://github.com/search?q=org%3Anf-core+path%3A*.nf+args%5B0+&type=code).

I agree it clashes with our discourse on more formal input schemas.

I am in favour of dropping it.

ewels commented:
I've heard of some folks using it, eg. @felixkrueger / @s-andrews at Babraham.

We've actually just made use of it for the first time in @nf-core a few days ago, purely as a method to autodetect malformed `profile` statements (it's a common error to put a space after the comma, eg. `-profile test, docker`).

Agree that we should remove it, as the majority of the time it's just a way for stuff to fail silently. Having a hard failure would be better.

The main valid use is to be able to glob input files. But yeah, I think very few do that and there are probably better ways to handle it (eg. if / when we build tooling to automate the generation of sample sheets for local files).

bentsherman committed:
Mergre upstream changes

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
cleanup

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman committed:
Remove positional args from v2 run, inspect

Signed-off-by: Ben Sherman <bentshermann@gmail.com>

bentsherman commented:
From our conversation about this PR last week -- if we remove run args from CLI v1, that would allow us to migrate CLI v1 from jcommander to picocli, with the benefits of better 
parsing and help text. It is one way to salvage PR if we don't have the appetite to maintain a separate CLI just to use double dashes for long options. But it might just be a good thing to do either way.

My understanding was that if we do release CLI v2 then we would eventually deprecate the old CLI, so the extra scaffolding to maintain both would be temporary.

bentsherman commented:
Closing until we choose to pick this up again
